### PR TITLE
database: re-add UserPublicRepoStore

### PR DESCRIPTION
**Pull request by arussellsaw**
This PR re-adds the user_public_repos table that i deleted previously, as we're actually using it now, this PR is a precursor to a graphql endpoint that fills this table & clones any non-cloned repos.

### CODE REVIEW

File: internal/database/user_public_repos.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		"INSERT INTO user_public_repos(user_id, repo_uri, repo_id) VALUES %s",
+		sqlf.Join(values, ","),
+	))
+}
+
+func (s *UserPublicRepoStore) SetUserRepo(ctx context.Context, upr UserPublicRepo) error {
+	return s.store.Exec(ctx, sqlf.Sprintf(
+		`INSERT INTO
+			user_public_repos(user_id, repo_uri, repo_id)
+		VALUES (%v, %v,  %v)
```
Comment (unknwon) [inline internal/database/user_public_repos.go:64]: nit: why suddenly `%v` now?

File: migrations/frontend/1528395792_add_user_public_repos.up.sql [hunk 2]
```diff
@@ -0,0 +1,12 @@
+BEGIN;
+
+CREATE TABLE IF NOT EXISTS user_public_repos (
+    user_id integer NOT NULL,
+    repo_uri text NOT NULL,
+    repo_id integer NOT NULL,
+    UNIQUE(user_id, repo_uri),
```
Comment (unknwon) [inline migrations/frontend/1528395792_add_user_public_repos.up.sql:7]: What's the reason to not use `UNIQUE(user_id, repo_id)`? 🤔 Have unique constraint on two ID-keys seems making more sense?

**Review by unknwon — APPROVED** [review]

File: internal/database/user_public_repos.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+			userID, repo.RepoURI, repo.RepoID,
+		))
+	}
+	return tx.Exec(ctx, sqlf.Sprintf(
+		"INSERT INTO user_public_repos(user_id, repo_uri, repo_id) VALUES %s",
+		sqlf.Join(values, ","),
+	))
+}
+
+func (s *UserPublicRepoStore) SetUserRepo(ctx context.Context, upr UserPublicRepo) error {
```
Comment (ryanslade) [inline internal/database/user_public_repos.go:60]: Maybe add a comment here mentioning that it resolves conflicts?

File: internal/database/user_public_repos_test.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
+package database
+
+import (
+	"context"
+	"testing"
+
+	"github.com/sourcegraph/sourcegraph/internal/api"
+
```
Comment (ryanslade) [inline internal/database/user_public_repos_test.go:8]: nit: Empty line

**Review by ryanslade — APPROVED** [review]

File: internal/database/user_public_repos.go (hunk 1 above)
Comment (arussellsaw) [reply internal/database/user_public_repos.go:64]: i don't think i really thought about it, i seem to use %s and %v interchangeably based on what my muscle memory decides at the time 😅  i'll make em consistent

File: migrations/frontend/1528395792_add_user_public_repos.up.sql (hunk 2 above)
Comment (arussellsaw) [reply migrations/frontend/1528395792_add_user_public_repos.up.sql:7]: hmm, originally it was possible for repo_id to be null which is why i added it this way initially.

i figure using the user-entered value here made more sense as we will also be returning this list to the user, so having possible duplicates in the list of repo URIs that they submitted would be confusing...

i don't think it really matters though, i'll change it to repo_id

Comment (unknwon) [reply migrations/frontend/1528395792_add_user_public_repos.up.sql]: >i figure using the user-entered value here made more sense as we will also be returning this list to the user,

Ah, this brings up an unrelated question to this PR. What happens if user enters a public repo that has not yet been added to SG before? I know we could add that repo real quick but do we wait/hang for adding it? (how that process looks like)

Comment (arussellsaw) [reply migrations/frontend/1528395792_add_user_public_repos.up.sql]: that's coming in the PR queued up beind this one 😄

**Pull request merged**